pkg/runtime/controller.go

```diff
@@ -0,0 +1,399 @@
+
+	// new handler table is created for every config change. It uses handler factory
+	// to create new handlers.
+	ht := newHandlerTable(instanceConfig, handlerConfig,
+		func(handler *cpb.Handler, instances []*cpb.Instance) (adapter.Handler, error) {
+			return hb.Build(handler, instances, newEnv(handler.Name, c.handlerPool))
+		},
+	)
+
+	// current consistent view of the rules keyed by Namespace and then Name.
+	// ht (handlerTable) keeps track of handler-instance association.
+	ruleConfig := c.processRules(handlerConfig, instanceConfig, ht)
+
+	// Initialize handlers that are used in the configuration.
+	// Some handlers may not initialize due to errors.
+	ht.Initialize(c.table)
+
+	// Combine rules with the handler table.
+	// Actions referring to handlers in error, will be purged.
+	combineRulesHandlers(ruleConfig, ht.table)
+
+	// create rules that are in the format that resolver needs.
+	rules := convertToRuntimeRules(ruleConfig)
+
+	// Create new resolver and cleanup the old resolver.
+	c.resolverID++
+	resolver := newResolver(c.eval, c.identityAttribute, c.defaultConfigNamespace, rules, c.resolverID)
+	c.dispatcher.ChangeResolver(resolver)
+
+	// copy old for deletion.
+	oldTable := c.table
+	oldResolver := c.resolver
+
+	// set new
+	c.table = ht.table
+	c.resolver = resolver
+
+	// synchronous call to cleanup.
+	cleanupResolver(oldResolver, oldTable, maxCleanupDuration)
+
+	glog.Infof("Published snapshot with %d config items", len(c.configState))
+}
+
+// maxCleanupDuration is the maximum amount of time cleanup operation will wait
+// before resolver ref count does to 0. It will return after this duration without
+// calling Close() on handlers.
+const maxCleanupDuration = 10 * time.Second
+
+// watchChanges watches for changes and publishes a new resolver.
+// watchChanges is started in a goroutine.
+func watchChanges(wch <-chan store.Event, applyEvents applyEventsFn) {
+	// consume changes and apply them to data indefinitely
+	var timeChan <-chan time.Time
+	var timer *time.Timer
+	events := make([]*store.Event, 0, maxEvents)
+
+	for {
+		select {
+		case ev := <-wch:
+			if len(events) == 0 {
+				timer = time.NewTimer(time.Second)
+				timeChan = timer.C
+			}
+			events = append(events, &ev)
+		case <-timeChan:
+			timer.Stop()
+			timeChan = nil
+			glog.Info("Publishing %d events", len(events))
+			applyEvents(events)
+			events = events[:0]
+		}
+	}
+}
+
+// applyEvents applies given events to config state and then publishes a snapshot.
+func (c *Controller) applyEvents(events []*store.Event) {
+	for _, ev := range events {
+		switch ev.Type {
+		case store.Update:
+			c.configState[ev.Key] = ev.Value
+		case store.Delete:
+			delete(c.configState, ev.Key)
+		}
+	}
+	c.publishSnapShot()
+}
+
+// filterInstanceConfig filters the configState and produces instanceConfig
+// that points to valid templates.
+func (c *Controller) filterInstanceConfig() map[string]*cpb.Instance {
```

Comment:
IMO, this mtd can be called getInstances(). Filter seems more powerful than just ignoring bad reference instances. Same argument for similar other mtds